src/TierRegistry.sol

```diff
@@ -0,0 +1,330 @@
+     */
+    function batchPurchaseTiers(
+        uint256[] calldata fids,
+        uint256[] calldata tiers,
+        uint256[] calldata forDays,
+        address payer
+    ) external whenNotPaused {
+        if (fids.length == 0) revert InvalidBatchInput();
+        if (fids.length != tiers.length) revert InvalidBatchInput();
+        if (fids.length != forDays.length) revert InvalidBatchInput();
+
+        // Effects
+        uint256 totalCost;
+        for (uint256 i; i < fids.length; ++i) {
+            uint256 pricePerDay = tokenPricePerDay[tiers[i]];
+            uint256 numDays = forDays[i];
+            if (numDays == 0) revert InvalidAmount();
+            if (pricePerDay == 0) revert InvalidTier();
+            if (numDays < minDays) revert InvalidAmount();
+            if (numDays > maxDays) revert InvalidAmount();
+            totalCost += pricePerDay * numDays;
+        }
+
+        for (uint256 i; i < fids.length; ++i) {
+            emit PurchasedTier(fids[i], tiers[i], forDays[i]);
+        }
+
+        IERC20 token = IERC20(paymentToken);
+        token.safeTransferFrom(payer, vault, totalCost);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                         PERMISSIONED ACTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @inheritdoc ITierRegistry
+     */
+    function creditTier(uint256 fid, uint256 tier, uint256 forDays) external onlyOperator whenNotPaused {
+        if (forDays == 0) revert InvalidAmount();
+        if (tokenPricePerDay[tier] == 0) revert InvalidTier();
+        if (forDays < minDays) revert InvalidAmount();
+        if (forDays > maxDays) revert InvalidAmount();
+
+        emit PurchasedTier(fid, tier, forDays);
+    }
+
+    /**
+     * @inheritdoc ITierRegistry
+     */
+    function batchCreditTiers(
+        uint256[] calldata fids,
+        uint256[] calldata tiers,
+        uint256[] calldata forDays
+    ) external onlyOperator whenNotPaused {
+        // Pre-checks
+        if (fids.length == 0) revert InvalidBatchInput();
+        if (fids.length != tiers.length) revert InvalidBatchInput();
+        if (fids.length != forDays.length) revert InvalidBatchInput();
+
+        for (uint256 i; i < fids.length; ++i) {
+            if (forDays[i] == 0) revert InvalidAmount();
+            if (tokenPricePerDay[tiers[i]] == 0) revert InvalidTier();
+            if (forDays[i] < minDays) revert InvalidAmount();
+            if (forDays[i] > maxDays) revert InvalidAmount();
+        }
+
+        for (uint256 i; i < fids.length; ++i) {
+            emit PurchasedTier(fids[i], tiers[i], forDays[i]);
+        }
+    }
+
+    /**
+     * @inheritdoc ITierRegistry
+     */
+    function setVault(
+        address vaultAddr
+    ) external onlyOwner {
+        if (vaultAddr == address(0)) revert InvalidAddress();
+        emit SetVault(vault, vaultAddr);
+        vault = vaultAddr;
+    }
+
+    function setToken(
```

Comment:
One thing to be wary of with the flexibility to set token is that we need to set prices in accordance with the value of the token. The contract doesn't support translating a usd based price to a price in the token. I think this makes the flexibility here much less useful, but it's potentially worth having if we want to change the token in a pinch.